### PR TITLE
Better CPIX 2.2 support + other fixes/enhancements

### PR DESCRIPTION
**Pull request by Heronyme**
Summary of changes:

- Synchronized schema with official CPIX 2.2. XSD + key period filter IDREF bugfix from CPIX 2.3 schema.
- Made content key data optional in loaded documents (to align with CPIX spec). Still required on save.
- Added root ContentID attribute support.
- Added content key period support (only basic validation).
- Added partial key period filter support (read/write; but no content key resolution or advanced constraint checking)
- Added URIExtXKey support (this is deprecated, but has popular demand).
- Documents are now saved without BOM.
- Readme update.

### CODE REVIEW

File: Cpix/Constants.cs [hunk 1]
```diff
@@ -44,12 +44,12 @@ static class Constants
 		/// 
 		/// Values are name-namespace pairs.
 		/// </summary>
-		public static readonly Tuple<string, string>[] TopLevelXmlElementOrder = new Tuple<string, string>[]
+		public static readonly Tuple<string, string>[] TopLevelXmlElementOrder =
```
Comment (MichaelChristopherRooney) [inline Cpix/Constants.cs:47]: I can't highlight the exact line for some reason, but at line 41 "oder" should be "order".

File: Cpix/ContentKeyPeriod.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		{
+			ValidateEntity();
+		}
+
+		private void ValidateEntity()
+		{
+			if ((Index == null && (Start == null || End == null))
+				|| (Index != null && (Start != null || End != null)))
+			{
+				throw new InvalidCpixDataException("For each content key period either the index or both the start and end time must be specified.");
```
Comment (MichaelChristopherRooney) [inline Cpix/ContentKeyPeriod.cs]: I would reword to "A content key period must specify either only the index or both the start and end time,"

File: Tests/SmokeTests.cs [hunk 3]
```diff
@@ -78,6 +78,25 @@ public void Save_OneEncryptedKey_DoesNotHorriblyFail()
...
 		}
 
+		[Fact]
+		public void Save_DoesNotAddBomToStream()
+		{
+			var expectedFirstFiveBytes = new [] { (byte)'<', (byte)'?', (byte)'x', (byte)'m', (byte)'l' } ;
+
+			var document = new CpixDocument();
+
+			using (var buffer = new MemoryStream())
```
Comment (KristjanJ) [inline Tests/SmokeTests.cs:90]: Maybe just me, but a variable named 'buffer' is unusual for a stream.

File: Cpix/ContentKeyPeriod.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+		internal override void ValidateLoadedEntity(CpixDocument document)
+		{
+			ValidateEntity();
+		}
+
+		private void ValidateEntity()
+		{
+			if ((Index == null && (Start == null || End == null))
+				|| (Index != null && (Start != null || End != null)))
```
Comment (KristjanJ) [inline Cpix/ContentKeyPeriod.cs]: Can this be made easier to understand?

File: Cpix/ContentKeyPeriod.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+		/// with start and end.
+		/// </summary>
+		public int? Index { get; set; }
+
+		/// <summary>
+		/// Gets or sets the wall clock (Live) or media time (VOD) for the start time
+		/// for the period. Mutually inclusive with end, and mutually exclusive with
+		/// index.
+		/// </summary>
+		public DateTime? Start { get; set; }
```
Comment (KristjanJ) [inline Cpix/ContentKeyPeriod.cs]: Is there a particular reason to use DateTime? DateTimeOffset is considered the more modern alternative.

DateTime has all kinds of edge cases and one would always need to convert all DateTimes to UTC to compare them. If the kind of a DateTime is not known, then comparisons will break.

For exampe:
`var dt1 = DateTime.Now;
var dt2 = DateTime.UtcNow;
Console.WriteLine(dt1 - dt2); // Almost 3 hours difference.`

With DateTimeOffset such ambiguities cannot occur.

File: Readme.md [hunk 6]
```diff
@@ -1,7 +1,7 @@
 Axinom CPIX library
 ===================
 
-A .NET library for working with CPIX (Content Protection Information Exchange) documents, as defined by [DASH Industry Forum](http://dashif.org/guidelines/). This library implements CPIX version 2.1.
+A .NET library for working with CPIX (Content Protection Information Exchange) documents, as defined by [DASH Industry Forum](http://dashif.org/guidelines/). This library implements CPIX version 2.2.
```
Comment (KristjanJ) [inline Readme.md]: Isn't it .NET standard? If yes, perhaps valuable to mention. People associated .NET with the old stuff.

**Review by KristjanJ — COMMENTED** [review]
Reviewed.

File: Cpix/ContentKeyPeriod.cs (hunk 2 above)
Comment (sakvith) [reply Cpix/ContentKeyPeriod.cs]: Yeah, "either" pat makes me confuse too a bit.  Would go with Michael's.

File: Tests/ContentKeyPeriodCrudTests.cs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+			document.ContentKeyPeriods.Add(new ContentKeyPeriod { Index = expectedPeriod1Index });
+			document.ContentKeyPeriods.Add(new ContentKeyPeriod { Id = expectedPeriod2Id, Index = expectedPeriod2Index});
+			document.ContentKeyPeriods.Add(new ContentKeyPeriod { Id = expectedPeriod3Id, Start = expectedPeriod3Start, End = expectedPeriod3End });
+			document.ContentKeyPeriods.AddSignature(TestHelpers.Certificate1WithPrivateKey);
+
+			document = TestHelpers.Reload(document);
+
+			Assert.Equal(3, document.ContentKeyPeriods.Count);
+
+			Assert.Null(document.ContentKeyPeriods.ElementAt(0).Id);
```
Comment (sakvith) [inline Tests/ContentKeyPeriodCrudTests.cs:224]: Usually, we add a massage. Should it apply here? Or it's something we use internally.

File: Cpix/ContentKeyPeriod.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		}
+
+		internal override void ValidateLoadedEntity(CpixDocument document)
+		{
+			ValidateEntity();
+		}
+
+		private void ValidateEntity()
+		{
+			if ((Index == null && (Start == null || End == null))
```
Comment (MichaelChristopherRooney) [inline Cpix/ContentKeyPeriod.cs]: As discussed on Slack, this could be moved to make it clearer. Ex:

var invalid =
        (Index == null && (Start == null || End == null)) ||
	(Index != null && (Start != null || End != null));

if (invalid)
      .....

File: Cpix/ContentKeyPeriodCollection.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	{
+		public const string ContainerXmlElementName = "ContentKeyPeriodList";
+
+		public ContentKeyPeriodCollection(CpixDocument document) : base(document)
+		{
+		}
+
+		internal override string ContainerName => ContainerXmlElementName;
+
+		protected override XmlElement SerializeEntity(XmlDocument document, XmlNamespaceManager namespaces, XmlElement container, ContentKeyPeriod entity)
```
Comment (MichaelChristopherRooney) [inline Cpix/ContentKeyPeriodCollection.cs:16]: I'm a little confused about this. This class name contains "Collection", but it seems like only a single entity is serialized here? Perhaps some comments could help? Same applies to Deserialise below.

File: Cpix/CpixDocument.cs [hunk 10]
```diff
@@ -199,7 +237,7 @@ public void Save(Stream stream)
 				{
 					// NB! Do not apply any formatting here, as digital signatures have already been generated
 					// and any formatting will invalidate the signatures!
-					Encoding = Encoding.UTF8,
+					Encoding = new UTF8Encoding(false),
```
Comment (MichaelChristopherRooney) [inline Cpix/CpixDocument.cs]: Perhaps a comment that "false" is used to stop a BOM from being emitted?

File: Cpix/Internal/XmlElements.cs [hunk 11]
```diff
@@ -213,6 +230,46 @@ public sealed class HlsSignalingDataElement
...
+			set => Index = value != null ? (int?)int.Parse(value) : null;
+		}
+
+		[XmlIgnore]
+		public DateTime? Start { get; set; }
+
+		[XmlAttribute("start")]
+		public string StartAsXmlString
+		{
+			get => Start?.ToString("o");
```
Comment (MichaelChristopherRooney) [inline Cpix/Internal/XmlElements.cs]: What's the significance of "o" here? Perhaps a comment can explain

File: Cpix/UsageRule.cs [hunk 12]
```diff
@@ -36,6 +37,14 @@ internal override void ValidateLoadedEntity(CpixDocument document)
 			if (!document.ContentKeys.Any(ck => ck.Id == KeyId))
 				throw new InvalidCpixDataException("Content key usage rule references a content key that is not present in the CPIX document.");
 
+			foreach (var keyPeriodFilter in KeyPeriodFilters)
+			{
+				keyPeriodFilter.Validate();
+
+				if (!document.ContentKeyPeriods.Any(ckp => ckp.Id == keyPeriodFilter.PeriodId))
+					throw new InvalidCpixDataException("Content key usage rule key period filter references a content key period that is not present in the CPIX document.");
```
Comment (MichaelChristopherRooney) [inline Cpix/UsageRule.cs]: String is too long for one line,

File: Tests/ContentKeyPeriodCrudTests.cs [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+
+		[Fact]
+		public void AddContentKeyPeriod_WithVariousValidData_Succeeds()
+		{
+			var document = new CpixDocument();
+
+			Assert.Null(Record.Exception(() => document.ContentKeyPeriods.Add(new ContentKeyPeriod
+			{
+				Index = 1
+			})));
```
Comment (MichaelChristopherRooney) [inline Tests/ContentKeyPeriodCrudTests.cs:45]: newline after this? Plus below and in other test methods.

File: Tests/ContentKeyPeriodCrudTests.cs [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+
+			Assert.Equal(2, document.ContentKeyPeriods.Count);
+		}
+
+		[Fact]
+		public void AddContentKeyPeriod_WithVariousValidData_Succeeds()
+		{
+			var document = new CpixDocument();
+
+			Assert.Null(Record.Exception(() => document.ContentKeyPeriods.Add(new ContentKeyPeriod
```
Comment (MichaelChristopherRooney) [inline Tests/ContentKeyPeriodCrudTests.cs:42]: This seems to be a weird way to do this? Don't we usually use Assert.DoesNotThrow?

Comment (MichaelChristopherRooney) [reply Tests/ContentKeyPeriodCrudTests.cs:42]: Oh, is this Xunit? Perhaps that's why.

File: Tests/ContentKeyPeriodCrudTests.cs [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+			{
+				Id = "test",
+				Start = DateTime.Now,
+				End = DateTime.Now
+			})));
+		}
+
+		[Fact]
+		public void AddContentKeyPeriod_WithVariousInvalidData_Fails()
```
Comment (MichaelChristopherRooney) [inline Tests/ContentKeyPeriodCrudTests.cs:60]: Maybe more suitable to use Test Cases (or Xunit equivalent) here?

File: Tests/SmokeTests.cs (hunk 3 above)
Comment (Heronyme) [reply Tests/SmokeTests.cs:90]: I'd probably specify that it's some kind of stream in the name. But this naming is used throughout the lib, so I'll keep it at this time.

File: Cpix/ContentKeyPeriod.cs (hunk 4 above)
Comment (Heronyme) [reply Cpix/ContentKeyPeriod.cs]: I'll just reformat it according to Michael's suggestions.

File: Cpix/ContentKeyPeriod.cs (hunk 5 above)
Comment (Heronyme) [reply Cpix/ContentKeyPeriod.cs]: DateTime was used to accommodate "unspecified" timezone, which the XML datetime allows for and CPIX spec also make no mention that a timezone should be required. So, DateTime felt as the better fit ... in order not to lose any info (DateTimeOffset doesn't support an "unspecified" and considers it implictly as Local timezone).  

But upon further discussions, it seems that allowing unspecified timezones might be a recipe for confusion anyway and it might be better to consider all unspecified timezones as UTC -- it's at least constant, so handling doesn't depend any way on where the lib runs (different local timezone). Alternative would've been to consider input without timezone invalid outright -- but it's a trade-off between scaring users away with error messages or accepting some default.

So, I guess this is what we'll do. And in this case DateTimeOffset sounds fine. We can revisit this when we get more clarity from CPIX developers.

File: Tests/ContentKeyPeriodCrudTests.cs (hunk 7 above)
Comment (Heronyme) [reply Tests/ContentKeyPeriodCrudTests.cs:224]: Good idea, but since none of the existing tests use messages either, I'm gonna keep as is.

File: Cpix/CpixDocument.cs (hunk 10 above)
Comment (Heronyme) [reply Cpix/CpixDocument.cs]: Used this style:

Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier:false)

File: Cpix/ContentKeyPeriodCollection.cs (hunk 9 above)
Comment (MichaelChristopherRooney) [reply Cpix/ContentKeyPeriodCollection.cs:16]: @Heronyme could you comment on this and then I'll approve.

Comment (Heronyme) [reply Cpix/ContentKeyPeriodCollection.cs:16]: There can be many entities, these methods are call for each entity/element within the collection/xml. The collection in this case represents the "ContentKeyPeriodList" XML element, which can contain many "ContentKeyPeriod" elements. This class can also contain collection-level validation, if needed. Does this make sense?

Comment (MichaelChristopherRooney) [reply Cpix/ContentKeyPeriodCollection.cs:16]: 👍

**Review by MichaelChristopherRooney — APPROVED** [review]

File: Cpix/Internal/XmlElements.cs (hunk 11 above)
Comment (Heronyme) [reply Cpix/Internal/XmlElements.cs]: This converts the time to the standard "ISO 8601" format. This complies with the XSD "datetime" format and is otherwise also widely used.

File: Cpix/UsageRule.cs (hunk 12 above)
Comment (Heronyme) [reply Cpix/UsageRule.cs]: Just got shorter.

File: Tests/ContentKeyPeriodCrudTests.cs (hunk 13 above)
Comment (Heronyme) [reply Tests/ContentKeyPeriodCrudTests.cs:45]: Gonna keep atm. This is like this everywhere.

File: Tests/ContentKeyPeriodCrudTests.cs (hunk 15 above)
Comment (Heronyme) [reply Tests/ContentKeyPeriodCrudTests.cs:60]: Gonna keep atm. This style is used elsewhere as well.

**Review by KristjanJ — APPROVED** [review]
